domain/application/state/containerimagemetadata.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// ContainerImageMetadataState describes retrieval and persistence methods for storage.
+type ContainerImageMetadataState struct {
+	*domain.StateBase
+	logger logger.Logger
+}
+
+// NewContainerImageMetadataState returns a new state reference.
+func NewContainerImageMetadataState(factory database.TxnRunnerFactory, logger logger.Logger) *ContainerImageMetadataState {
+	return &ContainerImageMetadataState{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// Get returns the container image metadata with the given UUID.
+// applicationerrors.ResourceNotFound is returned if the UUID is not in the
+// table
+func (s *ContainerImageMetadataState) Get(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+) (application.ContainerImageMetadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Capture(err)
+	}
+
+	storageKey := containerImageMetadataStorageKey{StorageKey: resourceUUID.String()}
+	m := containerImageMetadata{}
+	stmt, err := s.Prepare(`
+SELECT &containerImageMetadata.*
+FROM   resource_oci_image_metadata_store
+WHERE  storage_key = $containerImageMetadataStorageKey.storage_key
+`, storageKey, m)
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf(
+			"preparing select OCI image resource metadata statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, storageKey).Get(&m)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ResourceNotFound
+		} else if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf("selecting OCI image metadata: %w", err)
+	}
+
+	return application.ContainerImageMetadata{
+		UUID:         containermetadataresource.UUID(m.UUID),
+		StorageKey:   m.StorageKey,
+		RegistryPath: m.RegistryPath,
+		Username:     m.UserName,
+		Password:     m.Password,
+	}, nil
+}
+
+// Put stores container image metadata the database and returns its UUID.
+func (s *ContainerImageMetadataState) Put(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+	registryPath, userName, password string,
+) (containermetadataresource.UUID, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	uuid, err := containermetadataresource.NewUUID()
+	if err != nil {
+		return "", errors.Errorf("generating new UUID: %w", err)
+	}
+
+	m := containerImageMetadata{
+		UUID:         uuid.String(),
+		StorageKey:   resourceUUID.String(),
+		RegistryPath: registryPath,
+		UserName:     userName,
+		Password:     password,
+	}
+
+	stmt, err := s.Prepare(`
+INSERT INTO resource_oci_image_metadata_store (*)
```

Comment:
**issue:** Usually Put means "insert or update". This implementation insert or fails. Either it is expected, and the naming is incorrect, either it is not expected, and the implementation is incorrect.

Reply:
Good point, this should update if the resource is already there. Will change